libraries/GCS_MAVLink/GCS_Common.cpp

```diff
@@ -431,48 +431,101 @@ void GCS_MAVLINK::send_ahrs3()
-    waypoint_receiving = false;             // record that we are sending commands (i.e. not receiving)
+    // reply with number of commands in the mission.  The GCS will
+    // then request each command separately
+    mavlink_msg_mission_count_send(chan,
+                                   msg->sysid,
+                                   msg->compid,
+                                   count,
+                                   packet.mission_type);
 }
 
 /*
   handle a MISSION_REQUEST mavlink packet
  */
-void GCS_MAVLINK::handle_mission_request(AP_Mission &mission, mavlink_message_t *msg)
+void GCS_MAVLINK::handle_mission_request_int(mavlink_message_t *msg)
 {
     AP_Mission::Mission_Command cmd;
 
-    if (msg->msgid == MAVLINK_MSG_ID_MISSION_REQUEST_INT) {  
         // decode
         mavlink_mission_request_int_t packet;
         mavlink_msg_mission_request_int_decode(msg, &packet);
 
+        mavlink_mission_item_int_t ret_packet{};
+
+        ret_packet.target_system = msg->sysid;
+        ret_packet.target_component = msg->compid;
+        ret_packet.seq = packet.seq;
+        ret_packet.mission_type = packet.mission_type;
+
+        MAV_MISSION_RESULT result_code;
+
+        switch (packet.mission_type) {
+        case MAV_MISSION_TYPE_MISSION:
+            result_code = handle_mission_request_int_mission(packet, ret_packet);
+            break;
+        default:
+            result_code = MAV_MISSION_UNSUPPORTED;
+            break;
+        }
+
+        if (result_code != MAV_MISSION_ACCEPTED) {
+            // send failure message
+            mavlink_msg_mission_ack_send(chan,
+                                         msg->sysid,
+                                         msg->compid,
+                                         result_code,
+                                         packet.mission_type);
+            return;
+        }
+
+        // we already have a filled structure, use it in place of _send:
+        _mav_finalize_message_chan_send(chan, 
+                                        MAVLINK_MSG_ID_MISSION_ITEM_INT,
+                                        (const char *)&ret_packet,
+                                        MAVLINK_MSG_ID_MISSION_ITEM_MIN_LEN,
+                                        MAVLINK_MSG_ID_MISSION_ITEM_INT_LEN,
+                                        MAVLINK_MSG_ID_MISSION_ITEM_INT_CRC);
+}
+
+MAV_MISSION_RESULT GCS_MAVLINK::handle_mission_request_int_mission(const mavlink_mission_request_int_t &packet, mavlink_mission_item_int_t &ret_packet)
+{
+    AP_Mission *mission = AP::mission();
+    if (mission == nullptr) {
+        return MAV_MISSION_UNSUPPORTED;
+    }
+
+    AP_Mission::Mission_Command cmd;
+
         // retrieve mission from eeprom
-        if (!mission.read_cmd_from_storage(packet.seq, cmd)) {
-            goto mission_item_send_failed;
+        if (!mission->read_cmd_from_storage(packet.seq, cmd)) {
+            return MAV_MISSION_ERROR;
```

Comment:
Technically the only way to get this is actually `INVALID_SEQUENCE`

Reply:
I think we shouldn't rely on that though. Currently we always assume that reading bytes from storage is successful but that's not necessarily true.